### ISSUE TITLE
Native MathML rendering problems

### ISSUE THREAD

**Issue by MurrayIII** [w3c/mathml-core#229]
Native MathML rendering differs from TeX/MathJax/Word in a number of ways. Here's a short list of differences.

| status | rendering problem |
| ------ | ---------------------------------------------------------------- |
| resolved | <​math display="block"> often doesn't center the equation |
|  | 𝑎²+𝑏² displays the superscript ² for the 𝑏 too high (should be at the same height as for the 𝑎). Also, a subscript for a character with a descender is too low. E.g., for 𝑎₂+𝑞₃, the ₃ is displayed too low (should have the same descent as the ₂) |
|  | Adjacent fractions need to be separated by a small space instead of being contiguous |
|  | 𝑎 sin 𝜃 needs a space between the 𝑎 and the sin 𝜃 and between the sin and the 𝜃 |
|  | &#x007C;x&#x007C; displays the vertical bars too high |
|  |  √𝜋 displays the square root too low |
|  | Parentheses around matrices are too large |
|  | Unicode spaces have incorrect widths, e.g., U+2009 (thin space) is a very thick space. U+2061 should have a thin-space width but has 0 width. | 
|  | The tilde combining mark U+303 should stretch over its base |
| resolved | In &#x003C;math display="block">, ∫'s should be larger as in TeX |
|  resolved | Displaying a fraction in block mode is centered, but the fraction bar length equals the display width instead of the width of the larger argument |

Here's a link to a document that illustrates the differences:
[Native MathML Display Bugs.docx](https://github.com/w3c/mathml-core/files/14875616/Native.MathML.Display.Bugs.docx)

**Comment by bkardell** [comment]
@MurrayIII can you clarify if you are speaking about a particular implementation? I assume you mean chromium?

**Comment by MurrayIII** [comment]
Yes, chromium since Edge uses chromium for rendering. You can see the problems using Chrome too. Try running https://murrayiii.github.io/UnicodeMathML/playground/ and type in some equations. In the settings sure that MathJax isn't the default rendering agent. For example, type in a^2+b^2=c^2 and see

[A black background with white numbers and symbols  Description automatically generated]

Or a/b and see

[A screenshot of a phone  Description automatically generated]

Thanks,
Murray

**Comment by MurrayIII** [comment]
Another issue is that a hat (^) accent over a math-italic f (𝑓) is displayed too far to the left. It's placed correctly for a math-italic x (𝑥), although it would be nice to use a smaller hat for the x.

**Comment by NSoiffer** [comment]
I have [created a codepen for these](https://codepen.io/nms/pen/RwOEMdW). I rearranged the examples to group them a bit based on similarities. Here's my take on them:

> <​math display="block"> often doesn't center the equation

I don't see this in the codepen. @MurrayIII can you come up with a reproducible example and create a codepen for it?

Maybe font related issues -- for these I created a table that uses Cambria Math and STIX Two Math (if you have them installed). Both should have good Open Type math tables, so any poor quality display is because of the spec or a bug in the implementation...
> 𝑎²+𝑏² displays the superscript ² for the 𝑏 too high (should be at the same height as for the 𝑎)

TeX uses a superscript shift that is a based on at least two conditions as does the [spec](https://w3c.github.io/mathml-core/#base-with-superscript). TeX's values were chosen so that in simple cases such as this, the scripts are aligned vertically. I haven't compared the spec to TeX, so I can't say whether the spec should be fixed or it is an implementation bug.
@fred-wang: do you remember if you tried to follow TeX's script rules when you wrote the spec algorithm?

> |x| displays the vertical bars too high

I added ()s to the example and they do the same thing. Firefox displays them well. I suspect this happens because `symmetric=true` is not being respected for non-stretchy versions of the characters.
The [spec](https://w3c.github.io/mathml-core/#layout-of-operators) is problem here:
> If it is not possible to [shape a stretchy glyph](https://w3c.github.io/mathml-core/#dfn-shape-a-stretchy-glyph) corresponding to c in the inline direction with the [first available font](https://www.w3.org/TR/css-fonts-4/#first-available-font) then fall back to the layout algorithm of [3.2.1.1 Layout of <mtext>](https://w3c.github.io/mathml-core/#layout-of-mtext).

I believe this is not correct. If the `mo` has `symmetric=true`, then that should still apply. I'll hold off opening an issue for that based on feedback to my claim this needs spec improvement.

> √𝜋 displays the square root too low

This might be another case of the non-stretched version being drawn wrong. Or it could be that the implementation is prioritizing the drawing so that the `RadicalGap` is set to the value from the OpenType math table. The [spec details are here](https://w3c.github.io/mathml-core/#square-root).
@fred-wang: do you remember whether you tried to emulate TeX with the radical layout rules?

> Parentheses around matrices are too large

This is the subject of #238 and would be solved by changing to using 90% of the content size.

> Adjacent fractions need to be separated by a small space instead of being contiguous
> 𝑎 sin 𝜃 needs a space between the 𝑎 and the sin 𝜃 and between the sin and the 𝜃

For both of these, I think the spec and implementation are correct. There should be U+2061 or U+2062 used (I added them to the codepen). On these characters, the `lspace` or `rspace` value should be set to be a thinspace (0.166em). I believe the default lspace/rspace rules get things normally right, but like TeX, there are places where you need to manually tweak them.

> Displaying a fraction in block mode is centered, but the fraction bar length equals the display width instead of the width of the larger argument.

I couldn't reproduce this.
@MurrayIII: can you come up with a codepen that reproduces this?

There is another issues: the interline spacing is too tight in Chrome. Here is a rendering in chrome that shows the interline gap is bad (seems to be 0):
![image](https://github.com/w3c/mathml-core/assets/1545836/3ecaafcf-ad5c-42ae-8e8b-53fb88cb2998)
This image was created using consecutive `math` elements (`display="block"`) with no tagging between them. In Firefox, the spacing looks good.

**Comment by khaledhosny** [comment]
> > Adjacent fractions need to be separated by a small space instead of being contiguous
> > 𝑎 sin 𝜃 needs a space between the 𝑎 and the sin 𝜃 and between the sin and the 𝜃
> 
> For both of these, I think the spec and implementation are correct. There should be U+2061 or U+2062 used (I added them to the codepen). On these characters, the `lspace` or `rspace` value should be set to be a thinspace (0.166em). I believe the default lspace/rspace rules get things normally right, but like TeX, there are places where you need to manually tweak them.

In TeX, `\sin` is a `\mathop`, so to get similar spacing in MathML, I think something like this would be needed (I don’t know, if it is proper MathML use, though):

```xml
<math display="block">
  <mrow>
    <mi>𝑎</mi>
    <mo>sin</mo>
    <mi>𝜃</mi>
  </mrow>      
</math>
```

<img width="134" alt="image" src="https://github.com/w3c/mathml-core/assets/93914/018ad580-772c-4ebd-a73e-b844d161dd9c">

**Comment by NSoiffer** [comment]
I agree with what you say about what happens in TeX, but in MathML `sin` should be in an `mi` -- see https://w3c.github.io/mathml/#presm_mi_examples.

The markup (as in the codepen) should be something like:
```
<math display="block">
  <mrow>
    <mi>𝑎</mi>
    <mo>&#x2062;</mo>
    <mrow>
      <mi>𝑠in</mi>
      <mo>&#x2061;</mo>
      <mi>𝜃</mi>
   </mrow>
  </mrow>
 </math>
```

I haven't thought through the consequences, but maybe the default for U+2061 (invisible function application) is wrong and should be to have a `lspace` or `rspace` of 0.166em. That would deal with one of the cases. U+2062 (invisible times) definitely should be have zero spacing (e.g, $2ab$).

**Comment by runarberg** [comment]
I was experiencing similar issues with the [mathup](https://runarberg.github.io/mathup/) authoring tool. I decided against inferring function application and invisible times when authors write e.g. `a sin theta`<sup>†</sup>, opting instead add inline start padding to the relevant elements. The CSS selector to target the correct elements was massive (the `:has()` selector was indeed very helpful here). If anyone is interested here is the relevant portion:

https://github.com/runarberg/mathup/blob/main/src/stylesheets/core.css#L1-L68

---

†: The reason being I cannot guarantee what authors mean so I would rather users of assistive technology be lied to. I still offer author a way to be implicit by providing zero-with operators `.$`, `.*`, `.+`, and `.,` for function application, invisible times, invisible add, and invisible separator respectively.

**Comment by orbita2d** [comment]
An issue touched upon above, which I think deserves another mention:
```xml
<math>
  <msub>
    <mover accent="true">
      <mi>H</mi>
      <mo>^</mo>
    </mover>
  </msub>
</math>
```
![capture2024-08-07T23:40:52](https://github.com/user-attachments/assets/e9b6b2d1-2672-48bd-a33a-bf41d701690c)


Hat symbols (latex \hat{...}) should be rendered as a circumflex. The above renders too wide and too bold.

```xml
<math>
  <msub>
    <mover accent="true">
      <mi>H</mi>
      <mo>&#x0302;</mo>
    </mover>
  </msub>
</math>
```
![capture2024-08-07T23:40:13](https://github.com/user-attachments/assets/ee443374-7316-4db9-a07e-5f57291212ab)

is better but still too wide.

Is it worth having accents like the above as first class citizens?

**Comment by khaledhosny** [comment]
```xml
<math>
  <msub>
    <mover accent="true">
      <mi>H</mi>
      <mo stretchy="false">&#x0302;</mo>
    </mover>
  </msub>
</math>
```

**Comment by MurrayIII** [comment]
The following two items in my original list were the result of my applying incorrect CSS styles to native math rendering:
1. <​math display="block"> often doesn't center the equation
2. Displaying a fraction in block mode is centered, but the fraction bar length equals the display width instead of the width of the larger argument
These items can be removed from this issue. The remaining items are still problems. Ideally the native rendering should match MathJax's (and LaTeX's).

**Comment by MurrayIII** [comment]
Re accents, the [OfficeMath](https://devblogs.microsoft.com/math-in-office/officemath/) display engine chooses the best accent width for the base. For example, different widths are used for i, m, a+b, etc. That way the user/MathML doesn't have to specify the width.

**Comment by MurrayIII** [comment]
Update the opening comment to place the rendering problems in a table with a leading column containing the current status, e.g., "resolved".

**Comment by bkardell** [comment]
I've added a new label "browser-tracking" to this because I believe it is not entirely possible to say these are spec probems or browser bugs, so I'd like to be able to filter them out when trying to look at spec-related issues

**Comment by MurrayIII** [comment]
Added two entries to the table at the start of this issue: Unicode spaces display with incorrect widths and the tilde combining mark should stretch over its base.

**Comment by fred-wang** [comment]
This report only mentions that Edge/Chromium is used, but is missing information like the ones I mention in #318. I can probably try to do that based on the playground website Murray provided, but at least having the UnicodeMath or LaTeX input would help. Also this is reporting multiple issues, we should probably split into individual issues as Neil as done for issue #238.

> @fred-wang: do you remember if you tried to follow TeX's script rules when you wrote the spec algorithm?
> @fred-wang: do you remember whether you tried to emulate TeX with the radical layout rules?

MathML Core tries to use the MATH table but unfortunately the OpenType spec is unclear. So it was a lot of guesses from existing Firefox's code (inspired by TeXBook rules but contains a lot of comments this cannot exactly match MathML) and from the "OpenType math illuminated" paper from Ulrik Vieth (about how to use the table in TeX engines) and other discussions with XeTeX / LuaTeX developers. It's possible some rules are missing or were badly interpreted. As usual, we also need to keep issue #78 in mind, as that may affect how vertical shifts and gaps are interpreted.

**Comment by fred-wang** [comment]
> > Adjacent fractions need to be separated by a small space instead of being contiguous
> > 𝑎 sin 𝜃 needs a space between the 𝑎 and the sin 𝜃 and between the sin and the 𝜃
> 
> For both of these, I think the spec and implementation are correct.

I've opened https://github.com/MurrayIII/UnicodeMathML/issues/13 for the adjacent fractions issue.
